### PR TITLE
Magpie adapter

### PR DESCRIPTION
**Pull request by dbyrns**

### CODE REVIEW

File: magpie/magpieadapter/__init__.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
+import logging
+import os
+import sys
+
+# TODO Is the following block really useful?
```
Comment (fmigneault) [inline magpie/magpieadapter/__init__.py]: this is changed in the branch I'm currently working on (api+unittests)
these imports are required to allow finding proper modules/files no matter where the $pwd is set

File: magpie/magpieadapter/__init__.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+
+from twitcher.adapter.base import AdapterInterface
+from twitcher.owsproxy import owsproxy
+
+
+logger = logging.getLogger(__name__)
+
+
+__version__ = '0.1.1'
```
Comment (fmigneault) [inline magpie/magpieadapter/__init__.py]: maybe return this in the API's `/version` route

File: magpie/magpieadapter/__init__.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        # TODO For magpie we cannot store the servicestore object since the constructor need a header with token
+        # taken from the request... maybe we should check for that?!?
+        #return MagpieOWSSecurity(tokenstore_factory(registry), servicestore_factory(registry))
+        return MagpieOWSSecurity()
+
+    def configurator_factory(self, settings):
+        from pyramid.config import Configurator
+        from pyramid.authentication import AuthTktAuthenticationPolicy
+        from pyramid.authorization import ACLAuthorizationPolicy
+        from ziggurat_foundations.models import groupfinder
```
Comment (fmigneault) [inline magpie/magpieadapter/__init__.py]: use 
`from magpie.definitions.ziggurat_definitions import groupfinder` 
`from magpie.definitions.ziggurat_definitions import Configurator, ...`

I would also create a similar 'twitcher_definitions' file to allow import everything from a single call/location

File: magpie/magpieadapter/__init__.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+            magpie_secret,
+            callback=groupfinder,
+        )
+        authz_policy = ACLAuthorizationPolicy()
+
+        config = Configurator(
+            settings=settings,
+            authentication_policy=authn_policy,
+            authorization_policy=authz_policy
+        )
```
Comment (fmigneault) [inline magpie/magpieadapter/__init__.py]: from `authn_policy` (L44) to here is identical to [magpie main L131-L142](https://github.com/Ouranosinc/Magpie/blob/5fde2f85743ce7cdbb2b0821d9be4cae777a99fd/magpie/magpie.py#L131-L142)
could be a nice little function defined in `security.py`

File: magpie/magpieadapter/magpieowssecurity.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
+import tempfile
+from twitcher.owsexceptions import OWSAccessForbidden
+from twitcher.utils import parse_service_name
+from twitcher.esgf import fetch_certificate, ESGF_CREDENTIALS
+from twitcher.datatype import Service
+from magpie.services import service_factory
+from magpie.models import Service
+from magpie.api.api_except import evaluate_call, verify_param
+from pyramid.httpexceptions import HTTPForbidden, HTTPNotFound
+from pyramid.interfaces import IAuthenticationPolicy, IAuthorizationPolicy
```
Comment (fmigneault) [inline magpie/magpieadapter/magpieowssecurity.py]: idem import 'definitions'

File: magpie/magpieadapter/magpieservice.py [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+import logging
+import requests
+import json
+LOGGER = logging.getLogger(__name__)
+
+
+from twitcher.store.base import ServiceStore
+from twitcher.datatype import Service
+from twitcher.exceptions import ServiceNotFound
+from pyramid.exceptions import ConfigurationError
```
Comment (fmigneault) [inline magpie/magpieadapter/magpieservice.py]: all `ServiceNotFound` / `ConfigurationError` could maybe also use `raise_http`

File: magpie/magpieadapter/magpieowssecurity.py [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+            #should contain all the acl, this the only thing important
+            permission_requested = service_specific.permission_requested() #parse request (GET/POST) to get the permission requested for that service
+
+            if permission_requested:
+                authn_policy = request.registry.queryUtility(IAuthenticationPolicy)
+                authz_policy = request.registry.queryUtility(IAuthorizationPolicy)
+                principals = authn_policy.effective_principals(request)
+                has_permission = authz_policy.permits(service_specific, principals, permission_requested)
+                if not has_permission:
+                    raise OWSAccessForbidden("Not authorized to access this resource.")
```
Comment (fmigneault) [inline magpie/magpieadapter/magpieowssecurity.py]: maybe could be `raise_http` ?
might need to investigate if it behaves as wanted for xml output (is it even required or a json would be possible?)
might need to have `OWSException` derive from `HTTPException` or `HTTPError` to allow checks to pass...

File: magpie/magpieadapter/postgresdb.py [hunk 8]
```diff
@@ -0,0 +1,72 @@
+from sqlalchemy import engine_from_config
```
Comment (fmigneault) [inline magpie/magpieadapter/postgresdb.py]: functions in this file are duplicate from `magpie.db`

File: magpie/magpieadapter/__init__.py (hunk 1 above)
Comment (dbyrns) [reply magpie/magpieadapter/__init__.py]: 👍

File: magpie/magpieadapter/__init__.py (hunk 2 above)
Comment (dbyrns) [reply magpie/magpieadapter/__init__.py]: Forgot to remove that... don't thing that the adapter needs its own version

**Comment by dbyrns** [comment]
For the import and reusability I would prefer that you take the lead if you don't mind. You know better Magpie. (Maybe create a branch from this one and PR what you have in mind?)

**Comment by fmigneault** [comment]
@dbyrns ok will try making a sub-branch/PR

**Pull request merged**